pallets/phala/src/basepool.rs

```diff
@@ -0,0 +1,645 @@
+				Origin::<T>::Signed(pallet_id()).into(),
+				Some(contributer),
+				cid,
+				None,
+				None,
+				Default::default(),
+				true,
+				None,
+			)?;
+
+			let attr = NftAttr { shares, stakes };
+			Self::set_nft_attr_inner(cid, nft_id, &attr)?;
+			pallet_rmrk_core::Pallet::<T>::set_lock((cid, nft_id), true);
+			Ok(nft_id)
+		}
+
+		#[frame_support::transactional]
+		pub fn burn_nft(
+			base_pool: &BasePool<T::AccountId, BalanceOf<T>>,
+			nft_id: NftId,
+		) -> DispatchResult {
+			Self::burn_nft_inner(base_pool.cid, nft_id)
+		}
+
+		#[frame_support::transactional]
+		pub(super) fn burn_nft_inner(cid: CollectionId, nft_id: NftId) -> DispatchResult {
+			pallet_rmrk_core::Pallet::<T>::set_lock((cid, nft_id), false);
+			pallet_rmrk_core::Pallet::<T>::burn_nft_by_issuer(
+				Origin::<T>::Signed(pallet_id()).into(),
+				cid,
+				nft_id,
+			)?;
+
+			Ok(())
+		}
+
+		/// Merge multiple nfts belong to one user in the pool.
+		#[frame_support::transactional]
+		pub fn merge_or_init_nft_for_staker(
+			base_pool: &BasePool<T::AccountId, BalanceOf<T>>,
+			staker: T::AccountId,
+		) -> Result<NftId, DispatchError> {
+			Self::merge_or_init_nft_for_staker_inner(base_pool.cid, staker)
+		}
+
+		pub(super) fn merge_or_init_nft_for_staker_inner(
+			cid: CollectionId,
+			staker: T::AccountId,
+		) -> Result<NftId, DispatchError> {
+			let nftid_arr: Vec<NftId> = pallet_rmrk_core::Nfts::<T>::iter_key_prefix(cid).collect();
+			let mut total_stakes: BalanceOf<T> = Zero::zero();
+			let mut total_shares: BalanceOf<T> = Zero::zero();
+			for nftid in &nftid_arr {
+				let nft = pallet_rmrk_core::Nfts::<T>::get(cid, nftid)
+					.ok_or(pallet_rmrk_core::Error::<T>::NoAvailableNftId)?;
+				if nft.owner
+					!= rmrk_traits::AccountIdOrCollectionNftTuple::AccountId(staker.clone())
+				{
+					continue;
+				}
+
+				let property = Self::get_nft_attr_inner(cid, *nftid)?;
+				total_stakes += property.stakes;
+				total_shares += property.shares;
+				Self::burn_nft_inner(cid, *nftid)?;
+			}
+
+			Self::mint_nft_inner(cid, staker, total_shares, total_stakes)
+		}
+
+		pub fn get_nft_attr(
+			base_pool: &BasePool<T::AccountId, BalanceOf<T>>,
+			nft_id: NftId,
+		) -> Result<NftAttr<BalanceOf<T>>, DispatchError> {
+			Self::get_nft_attr_inner(base_pool.cid, nft_id)
+		}
+
+		pub(super) fn get_nft_attr_inner(
+			cid: CollectionId,
+			nft_id: NftId,
+		) -> Result<NftAttr<BalanceOf<T>>, DispatchError> {
+			let key: BoundedVec<u8, <T as pallet_uniques::Config>::KeyLimit> = NFT_PROPERTY_KEY
+				.as_bytes()
+				.to_vec()
+				.try_into()
+				.expect("str coverts to bvec should never fail; qed.");
+			let raw_value: BoundedVec<u8, <T as pallet_uniques::Config>::ValueLimit> =
+				pallet_rmrk_core::Pallet::<T>::properties((cid, Some(nft_id), key))
+					.ok_or(Error::<T>::MissCollectionId)?;
+			Ok(Decode::decode(&mut raw_value.as_slice()).expect("Decode should never fail; qed."))
+		}
+
+		#[frame_support::transactional]
```

Comment:
When the inner function is already transactional, an additional transactional attribute only wastes cpu

Reply:
the transactional outside is removed